### PR TITLE
Fix update apply-on-update-default-values Add Feature (pressing OK)

### LINKED ISSUE #57411: Default values depending on other fields are not correctly applied when field not part of form made by Drag and Drop Designer
### What is the bug or the crash?

I have a vector layer with two fields: `field_a` and `field_b` and I have set up an attribute form using the Drag and Drop Designer. In my case, "field_a" is editable and has no default value. On the other hand, "field_b" is not editable, but has a default value definition in which "field_a" is used, e.g. like `title("field_a")`. 

Now, if "field_b" is not part of the attribute form set up using the drag and drop designer, it appears that the default value is not applied when adding or modifying features. In fact, the return value of `"field_a"` (as used in the default value expression) is always `NULL`. This does not happen if "field_b" is present in the form.

### Steps to reproduce the issue

Download the [sample_project.zip](https://github.com/qgis/QGIS/files/15294090/sample_project.zip)
1. Open the sample project from the attached geopackage
1. In the TOC select "test_layer_a" and digitize a new feature
1. For "field_a" type in e.g. "london", hit OK
1. Select "test_layer_b" repeat the prevvious steps

test_digit_form

### Versions

QGIS version | 3.36.2-Maidenhead | QGIS code revision | 6d250527
-- | -- | -- | --
Qt version | 5.15.13
Python version | 3.12.3
GDAL/OGR version | 3.8.5
PROJ version | 9.4.0
EPSG Registry database version | v11.004 (2024-02-24)
GEOS version | 3.12.1-CAPI-1.18.1
SQLite version | 3.45.1
PDAL version | 2.6.3
PostgreSQL client version | 16.2
SpatiaLite version | 5.1.0
QWT version | 6.2.0
QScintilla2 version | 2.14.1
OS version | Windows 11 Version 2009
  |   |   |  
Active Python plugins
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by signedav**
To have it aligned with the behavior on editing a feature.

This is compatible with the additional information from here #55633 and the fix here #56189 where volatile functions are not applied realtime but on pressing OK instead (on "Identify and open Form") - This behavior is now introduced for Add Feature as well, to have it aligned. 

This means:

#### Identify and open Form
- apply-on-update-default-values depending on fields are updated realtime in form
- apply-on-update-default-values are updated on OK (means volatile functions are relevant here, others do not change)

#### Add feature
- apply-on-update-default-values are set on form opens (like other default values)
- apply-on-update-default-values depending on fields are updated realtime in form
- apply-on-update-default-values are updated on OK (means volatile functions are relevant here, others do not change) *

\* but as well apply-on-update-default-values of invisible fields (fields that are not added by drag-n-drop) but depending on fields. This is basicly what this PR is about. So it fixes #57411

### CODE REVIEW

File: src/core/vector/qgsvectorlayer.cpp [hunk 1]
```diff
@@ -1373,6 +1373,11 @@ bool QgsVectorLayer::addFeature( QgsFeature &feature, Flags )
       success = mJoinBuffer->addFeature( feature );
   }
 
+  if ( success && !mDefaultValueOnUpdateFields.isEmpty() )
+  {
+    updateDefaultValues( feature.id() );
```
Comment (nirvn) [inline src/core/vector/qgsvectorlayer.cpp:1378]: @signedav , one thing I'm a tiny bit worried about here is the impact it'll have one default values that might rely on attribute form context and expression scopes (e.g. @current_parent). 

Could we do that on the attribute form side of things?

Comment (signedav) [reply src/core/vector/qgsvectorlayer.cpp:1378]: Thanks.

The reason why this is not done on attribute form side is, that (1) it should consider fields that are not added to the form via drag'n'drop as well and (2) to be aligned with the behavior on editing an existent feature. 

I understand the "weakness" that values could be changed twice (like a timestamp that is stored some seconds later than shown in the form before) but I don't understand what you mean with the @current_parent. Could you clarify the use case?

**Comment by signedav** [comment]
No stale. This should really be fixed in future. No matter if if on 3.38 or not.

**Comment by signedav** [comment]
Ping @nirvn

**Pull request closed without merging**